File: src/emulator/graphics/fetcher.nim
import ../[io, types, utils]
import bus
import deques
import bitops
import sdl2
import algorithm
import logging
import strformat

const DEFAULT_PALETTE = [sdl2.color(0xE0, 0xF8, 0xD0, 0xFF), #00 - WHITE
                        sdl2.color(0x88, 0xC0, 0x70, 0xFF), #01 - LIGHT GRAY
                        sdl2.color(0x34, 0x68, 0x56, 0xFF), #10 - DARK GRAY
                        sdl2.color(0x8, 0x18, 0x20, 0xFF)] #11 - BLACK

type
    Sprite* = ref object
        x*: uint8
        y*: uint8
        tileNum*: uint8
        flags*: uint8
        height*: uint16

    Pixel* = ref object
        colorCode*: uint8
        palette1*: bool
        priority*: bool

    Fetcher* = ref object
        state*: uint8
        spriteState*: uint8

        internalX*: uint16
        tileNum*: uint8
        tileDataAddr*: uint16
        tileLo*: uint8
        tileSlice: array[8, uint8]
        tFIFO* = initDeque[Pixel](8)

        isWindowVisible*: bool

        sprites*: seq[Sprite]
        spriteDataAddr*: uint16
        spriteLo*: uint8
        spriteSlice: array[8, uint8]
        currSprite*: Sprite
        sFIFO* = initDeque[Pixel](8)

        firstInstance*: bool
        tickingSprite*: bool

proc getTileDataOffset(tileNum: uint8): uint16 =
    if getLCDC(BGANDWINTILEDATAAREA):
        return getTileDataBase() + (uint16(tileNum) * 16)
    return getTileDataBase() + (uint16(cast[int8](tileNum)) * 16)

proc getTileRow(f: Fetcher): uint16 =
    if not f.isWindowVisible:
        return (((LY + SCY) mod 8).uint16 * 2)
    return (WLY mod 8).uint16 * 2

proc internalY(f: Fetcher): uint16 =
    if not f.isWindowVisible:
        return (((LY + SCY) and 0xFF).float / 8.0).uint16 * 32
    return (WLY.float / 8.0).uint16 * 32

proc getColorIndex*(code: uint8, obj: bool = false, palette1: bool = false): uint8 =
    let index = code.int * 2

    if obj:
        if palette1:
            return (OBP1.bitsliced(index..(index+1))).uint8
        return (OBP0.bitsliced(index..(index+1))).uint8
    return (BGP.bitsliced(index..(index+1))).uint8

proc searchColor*(index: uint8): Color = return DEFAULT_PALETTE[index]

proc fetchPixelSlice(f: Fetcher, lo, hi: uint8, sprite: bool = false): void =
    var 
        revX: int
        cc: uint8

    for x in countdown(7, 0):
        cc = (hi.testBit(x).uint8 shl 1) or lo.testBit(x).uint8
        if sprite: f.spriteSlice[revX] = cc
        else: f.tileSlice[revX] = cc

        inc revX

proc tick*(f: Fetcher): void =
    case f.state
    of 0:
        f.state = 1
    
    of 1:
        f.tileNum = readByte(getTileMapBase(win = f.isWindowVisible) + ((
                    f.internalX + f.internalY()) and 0x3FF))
        f.state = 2
    
    of 2:
        f.state = 3

    of 3:
        f.tileDataAddr = getTileDataOffset(f.tileNum) + f.getTileRow()
        f.tileLo = readByte(f.tileDataAddr)
        f.state = 4

    of 4:
        f.state = 5

    of 5:
        f.fetchPixelSlice(f.tileLo, readByte(f.tileDataAddr + 1))
        f.state = 6
        if f.firstInstance:
            f.state = 0
            f.firstInstance = false
            return

    of 6:
        if (f.tFIFO.len() != 0):
            return

        for i in 0..7:
            f.tFIFO.addLast(Pixel(colorCode: f.tileSlice[i]))

        f.internalX = (f.internalX + 1) mod 32
        f.state = 0

    else:
        logger.log(lvlError, fmt"TILE FETCHER REACHED AN UNKNOWN STATE. (state: {f.state})")

proc spriteTick*(f: Fetcher): void =
    case f.spriteState
    of 0:
        if f.currSprite.height == 16: f.currSprite.tileNum = (f.currSprite.tileNum and 0xFE)
        var yOff = (LY.uint16 - f.currSprite.y.uint16 + 16)
        if f.currSprite.flags.testBit(6): yOff = (f.currSprite.height - 1 - yOff)

        f.spriteDataAddr = 0x8000 + (uint16(f.currSprite.tileNum) * 16) + (yOff * 2)
        f.spriteLo = readByte(f.spriteDataAddr)
        f.spriteState = 1

    of 1:
        f.fetchPixelSlice(f.spriteLo, readByte(f.spriteDataAddr + 1), sprite=true)
        f.spriteState = 2

    of 2:
        let residualAmount = f.sFIFO.len()
        if f.currSprite.flags.testBit(5): f.spriteSlice.reverse()

        for i in 0..7:
            let
                cc = f.spriteSlice[i]
                px = Pixel(colorCode: cc, palette1: f.currSprite.flags.testBit(4), priority: f.currSprite.flags.testBit(7))

            if (residualAmount >= (i+1)):
                let residualPixel = f.sFIFO[i]
                if (residualPixel.colorCode == 0) and (cc != 0): f.sFIFO[i] = px

            else: f.sFIFO.addLast(px)

        f.sprites.delete(0)
        f.spriteState = 0
        f.tickingSprite = false

    else:
        logger.log(lvlError, fmt"SPRITE FETCHER REACHED AN UNKNOWN STATE. (state: {f.spriteState})")

proc fetch(f: Fetcher, xPos: uint16): void =
    f.state = 0
    f.internalX = xPos
    f.tFIFO.clear()

proc fetchBackground*(f: Fetcher): void =
    f.fetch((SCX.float / 8.0).uint16 and 0x1F)

proc fetchWindow*(f: Fetcher): void =
    var saturatedWX = WX - 7
    if WX <= 6: saturatedWX = 0
    f.fetch((((LX - saturatedWX)).float / 8.0).uint16 and 0x1F)

File: src/fetcher.c
#include "fetcher.h"

#include <SDL2/SDL.h>
#include <stdbool.h>
#include <stdint.h>

#include "fifo.h"
#include "io.h"
#include "logger.h"
#include "utils.h"
#include "vbus.h"

const SDL_Color DEFAULT_PALETTE[4] = {
    {0xFF, 0xF6, 0xD3, 0xFF},  // 00 - WHITE
    {0xF9, 0xA8, 0x75, 0xFF},  // 01 - LIGHT GRAY
    {0xEB, 0x6B, 0x6F, 0xFF},  // 10 - DARK GRAY
    {0x7C, 0x3F, 0x58, 0xFF}   // 11 - BLACK
};

Fetcher f = {0};
PixelFIFO tFIFO = {{.maxSize = 8}, {{0}}};
PixelFIFO sFIFO = {{.maxSize = 8}, {{0}}};
SpriteBuffer sprites = {{.maxSize = 10}, {{0}}};

uint16_t getTileDataOffset(uint8_t tileNum) {
    if (getLCDC(BGANDWINTILEDATAAREA)) {
        return getTileDataBase() + ((uint16_t)tileNum * 16);
    }
    return getTileDataBase() + ((uint16_t)((int8_t)tileNum) * 16);
}

uint16_t getTileRow(Fetcher *f) {
    if (!f->isWindowVisible) {
        return ((uint16_t)(f->staticLine % 8) * 2);
    }
    return (uint16_t)(ioRegs.WLY % 8) * 2;
}

uint16_t internalY(Fetcher *f) {
    if (!f->isWindowVisible) {
        return ((uint16_t)(((ioRegs.LY + ioRegs.SCY) & 0xFF) / 8) * 32);
    }
    return (uint16_t)(ioRegs.WLY / 8.0) * 32;
}

uint8_t getColorIndex(uint8_t code, bool obj, bool palette1) {
    int index = (int)code * 2;
    uint8_t palette;
    if (obj) {
        palette = palette1 ? ioRegs.OBP1 : ioRegs.OBP0;
    } else {
        palette = ioRegs.BGP;
    }
    return BEXTR(palette, index, index + 1);
}

SDL_Color searchColor(uint8_t colorIndex) {
    return DEFAULT_PALETTE[colorIndex];
}

void fetchPixelSlice(Fetcher *f, uint8_t lo, uint8_t hi, bool isSprite) {
    int revX = 0;
    uint8_t cc = 0;
    uint8_t *slice = isSprite ? f->spriteSlice : f->tileSlice;

    for (int x = 7; x >= 0; x--) {
        cc = (BT(hi, x) << 1) | BT(lo, x);
        slice[revX] = cc;
        revX++;
    }
}

void tileTick(Fetcher *f) {
    switch (f->fetcherState) {
        case 0:
            f->fetcherState = 1;
            break;

        case 1:
            f->staticLine = ioRegs.LY + ioRegs.SCY;
            f->tileNum =
                ppuReadByte(getTileMapBase(f->isWindowVisible) +
                                ((f->internalX + internalY(f)) & 0x3FF),
                            false);
            f->fetcherState = 2;
            break;

        case 2:
            f->fetcherState = 3;
            break;

        case 3:
            f->tileDataAddr = getTileDataOffset(f->tileNum) + getTileRow(f);
            f->tileLo = ppuReadByte(f->tileDataAddr, false);
            f->fetcherState = 4;
            break;

        case 4:
            f->fetcherState = 5;
            break;

        case 5:
            fetchPixelSlice(f, f->tileLo,
                            ppuReadByte(f->tileDataAddr + 1, false), false);
            f->fetcherState = 6;
            if (f->firstInstance) {
                f->fetcherState = 0;
                f->firstInstance = false;
                return;
            }
            break;

        case 6:
            if (FIFOLen(&tFIFO.base) != 0) return;

            for (int i = 0; i < 8; i++) {
                Pixel px = {f->tileSlice[i], false, false};
                pushEntry(tFIFO, px);
            }
            f->internalX = (f->internalX + 1) % 32;
            f->fetcherState = 0;
            break;

        default:
            logState(FATAL, "Invalid Tile Fetcher state.");
            break;
    }
}

void flipX(Fetcher *f) {
    int i = 7, j = 0;
    while (i > j) {
        uint8_t cc = f->spriteSlice[i];
        f->spriteSlice[i] = f->spriteSlice[j];
        f->spriteSlice[j] = cc;
        i--;
        j++;
    }
}

void spriteTick(Fetcher *f) {
    switch (f->spriteState) {
        case 0:
            // Ignore LSb of tall sprites.
            if (f->currSprite.height == 16) f->currSprite.tileNum &= 0xFE;

            uint16_t yOff =
                (uint16_t)ioRegs.LY - (uint16_t)f->currSprite.y + 16;

            // Vertical flip.
            if (BT(f->currSprite.flags, 6))
                yOff = f->currSprite.height - 1 - yOff;

            f->spriteDataAddr =
                0x8000 + ((uint16_t)f->currSprite.tileNum * 16) + (yOff * 2);
            f->spriteLo = ppuReadByte(f->spriteDataAddr, false);
            f->spriteState = 1;
            break;

        case 1:
            fetchPixelSlice(f, f->spriteLo,
                            ppuReadByte(f->spriteDataAddr + 1, false), true);
            f->spriteState = 2;
            break;
        case 2:
            // Horizontal flip.
            if (BT(f->currSprite.flags, 5)) flipX(f);

            // If the Sprite FIFO already has some residual pixels left then we
            // have to mix both the new and the residual pixel at position I
            // using a transparency check.
            uint8_t residualAmount = FIFOLen(&sFIFO.base);

            for (int i = 0; i < 8; i++) {
                uint8_t cc = f->spriteSlice[i];
                Pixel px = {cc, BT(f->currSprite.flags, 4),
                            BT(f->currSprite.flags, 7)};

                if (residualAmount >= (i + 1)) {
                    Pixel residualPixel = getEntryAt(sFIFO, &sFIFO, i);

                    if ((residualPixel.colorCode == 0) && cc != 0) {
                        setEntryAt(sFIFO, i, px);
                    }
                } else {
                    pushEntry(sFIFO, px);
                }
            }

            popEntry(sprites, &sprites);

            f->spriteState = 0;
            f->tickingSprite = false;
            break;

        default:
            logState(FATAL, "Invalid Tile Fetcher state.");
            break;
    }
}

void fetchMain(Fetcher *f, uint16_t xPos) {
    f->fetcherState = 0;
    f->internalX = xPos;
    clearFIFO(&tFIFO.base);
}

void fetchBackground(Fetcher *f) {
    fetchMain(f, ((uint16_t)(ioRegs.SCX / 8) & 0x1F));
}

void fetchWindow(Fetcher *f) {
    uint8_t saturatedWX = ioRegs.WX - 7;
    if (ioRegs.WX <= 6 && ioRegs.WX > 0) {
        saturatedWX = ioRegs.WX;
    }
    fetchMain(f, (uint16_t)((ioRegs.LX - saturatedWX) / 8) & 0x1F);
}

void clearSpriteFetcher() {
    clearFIFO(&sprites.base);
    clearFIFO(&sFIFO.base);
}
